scripts/publish-resolve-data.js

```diff
@@ -0,0 +1,218 @@
+// @ts-check
+
+/**
+ * This function is invoked by the publish.yml GitHub Action workflow and contains all of the dynamic logic needed
+ * for the various workflow trigger types. This avoids the need for the logic to be stored in fragile inline
+ * shell commands.
+ *
+ * @typedef {'--dry-run' | ''} DryRunFlag
+ *
+ * @typedef {{
+ *   version: string;
+ *   dry_run_flag: DryRunFlag;
+ *   success_comment: string;
+ *   publish_branch: string;
+ *   repo: string;
+ *   ref: string;
+ * }} PublishResolveData
+ *
+ * Partial from https://github.com/actions/toolkit/blob/c6b487124a61d7dc6c7bd6ea0208368af3513a6e/packages/github/src/context.ts
+ * @typedef {{
+ *   actor: string;
+ *   runId: number;
+ *   repo: { owner: string; repo: string };
+ * }} GitHubContext
+ *
+ * @param {{
+ *  github: import('octokit/dist-types').Octokit;
+ *  context: GitHubContext;
+ *  core: import('@actions/core');
+ * }} param
+ */
+module.exports = async ({ github, context, core }) => {
+  const data = await getPublishResolveData({ github, context });
+
+  // Ensure that certain outputs are always set
+  if (!data.version) {
+    throw new Error('The "version" to release could not be determined');
+  }
+  if (!data.publish_branch) {
+    throw new Error('The "publish_branch" could not be determined');
+  }
+
+  // Set the outputs to be consumed in later steps
+  core.setOutput('version', data.version);
+  core.setOutput('dry_run_flag', data.dry_run_flag);
+  core.setOutput('success_comment', JSON.stringify(data.success_comment)); // Escape the multi-line string
+  core.setOutput('publish_branch', data.publish_branch);
+  core.setOutput('ref', data.ref);
+  core.setOutput('repo', data.repo);
+};
+
+/**
+ * @param {{
+ *  github: import('octokit/dist-types').Octokit;
+ *  context: GitHubContext;
+ * }} param
+ *
+ * @returns {Promise<PublishResolveData>}
+ */
+async function getPublishResolveData({ github, context }) {
+  // We use empty strings as default values so that we can let the `actions/checkout` action apply its default resolution
+  const DEFAULT_REF = '';
+  const DEFAULT_REPO = '';
+
+  /**
+   * "The short ref name of the branch or tag that triggered the workflow run. This value matches the branch or tag name shown
+   * on GitHub. For example, feature-branch-1."
+   *
+   * Source: https://docs.github.com/en/actions/learn-github-actions/variables#default-environment-variables
+   */
+  const refName = process.env.GITHUB_REF_NAME;
```

Comment:
Refactored to use the env var now that we know the JS copy of `github` does not match the GitHub Actions shell one, and explicitly checking for it not being set. I believe it should be able for all triggers, but we can relax this if it is not because we will be covered by the check above before setting the outputs